assembly/src/todo/mod.rs

```diff
@@ -0,0 +1,432 @@
+use core::borrow::Borrow;
+
+use vm_core::{code_blocks::CodeBlock, CodeBlockTable, Felt, FieldElement, Operation, Program};
+
+use crate::{
+    parsers::{self, Instruction, Node},
+    procedures::Procedure,
+    ModuleProvider, ProcedureAst, ProcedureId, ProgramAst,
+};
+
+pub struct Assembler {
+    module_provider: Box<dyn ModuleProvider>,
+    cb_table: CodeBlockTable,
+    locals: Vec<Procedure>,
+    blocks: Vec<Vec<CodeBlock>>,
+    spans: Vec<Vec<Operation>>,
+}
+
+impl Assembler {
+    pub fn new() -> Self {
+        Self {
+            module_provider: Box::new(()),
+            cb_table: CodeBlockTable::default(),
+            locals: Vec::new(),
+            blocks: Vec::new(),
+            spans: Vec::new(),
+        }
+    }
+
+    pub fn with_cb_table(mut self, cb_table: CodeBlockTable) -> Self {
+        self.cb_table = cb_table;
+        self
+    }
+
+    pub fn with_module_provider<P>(mut self, provider: P) -> Self
+    where
+        P: ModuleProvider + 'static,
+    {
+        self.module_provider = Box::new(provider);
+        self
+    }
+
+    pub fn into_inner(self) -> CodeBlockTable {
+        self.cb_table
+    }
+
+    pub fn compile<S>(&mut self, source: S) -> Program
+    where
+        S: AsRef<str>,
+    {
+        let source = source.as_ref();
+        let ast = ProgramAst::try_from(source).unwrap();
+
+        self.compile_ast(&ast)
```

Comment:
I would probably not create a separate `compile_ast()` function and just put the logic of this function here.

Reply:
Function removed